cr.swift

```diff
@@ -0,0 +1,41 @@
+//
+//  Netguru iOS code review task
+//
+
+class paymentViewController: UIViewController {
+    var delegate: PaymentViewControllerDelegate
+    let customView = PaymentView()
+    let payment: Payment?
+
+    func callbacks() {
+        view.backgroundColor = UIColor(displayP3Red: 1, green: 1, blue: 1, alpha: 1)
+        customView.didTapButton = {
+            if let payment = self.payment {
+                self.delegate.didFinishFlow(amount: payment.amount
+                                            currency: payment.currency)
+            }
+        }
+    }
+
+    func viewDidLoad() {
+        navigationController?.setNavigationBarHidden(false, animated: false)
+        customView.centerXAnchor.constraint(equalTo: view.centerXAnchor)
+        customView.centerYAnchor.constraint(equalTo: view.centerYAnchor)
+        view.addSubview(customView)
+        fetchPayment()
+        setupCallbacks()
+    }
+
+    internal let ViewModel: PaymentViewModel
+
+    func fetchPayment() {
+        customView.statusText = "Fetching data"
+        ApiClient.sharedInstance().fetchPayment { payment in
```

Comment:
there are few things to discuss in here.
Assuming that `fetchPayment` looks smt like that
``` swift
func fetchPayment(completion: @escaping (Payment?) -> Void)
```

First is that the return value might be nil, but at the same time we don't know what went wrong I would recommend packing it into a result type and adding custom error so that we can show explicit error in the UI.

Second important thing is the single instance, I don't think there is a need to keep just one instance of `ApiClient` and it will make our testing much harder. We can have an api client constrained by the protocol so that we can provide any instance or class matching the protocol we want. This will also allow us in future to use things like composers or decorators and make our codebase much easier to maintain.

Let's take a look in practice

``` swift
enum PaymentError: Error {
    case noData
}

protocol PaymentAPI {
    func fetchPayment(completion: @escaping (Result<Payment, PaymentError>) -> Void)
}

struct PaymentProvider: PaymentAPI {
    func fetchPayment(completion: @escaping (Result<Payment, PaymentError>) -> Void) {
        ApiClient.sharedInstance().fetchPayment { payment in
            if let payment = payment {
                completion(.success(payment))
            } else {
                completion(.failure(.noData))
            }
        }
    }
}
```

now we need to conform the `ApiClient` to our new protocol like this 
```swift
extension ApiClient: PaymentAPI { }
```

thanks to it we can now freely write init like this

``` swift
    init(paymentViewModel: PaymentViewModel = .init(),
         apiClient: PaymentAPI = ApiClient.sharedInstance()
    ) {
        self.viewModel = paymentViewModel
        self.apiClient = apiClient
        super.init(nibName: nil, bundle: nil)
    }
```

and work with the response like this

``` swift
apiClient.fetchPayment { completion in
            switch completion {
            case .failure(let error):
                print(error)
            case .success(let payment):
                DispatchQueue.main.async {
                    // UI changes
                }
            }
```

you can see `DispatchQueue.main.async` in the success block it's important to make updates to User Interface only on the main thread. Fortunately Xcode is good in catching ui updates from background thread so this should be fairly easy to get in future as well.